compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -379,25 +395,92 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
         goal: Goal<'tcx, (ty::Const<'tcx>, Ty<'tcx>)>,
     ) -> QueryResult<'tcx> {
         let (ct, ty) = goal.predicate;
-        let nested_goals = self.eq(goal.param_env, ct.ty(), ty)?;
-        self.evaluate_all_and_make_canonical_response(nested_goals)
+        self.eq(goal.param_env, ct.ty(), ty)?;
+        self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
     }
 }
 
 impl<'tcx> EvalCtxt<'_, 'tcx> {
-    // Recursively evaluates a list of goals to completion, returning the certainty
-    // of all of the goals.
-    fn evaluate_all(
-        &mut self,
-        mut goals: Vec<Goal<'tcx, ty::Predicate<'tcx>>>,
-    ) -> Result<Certainty, NoSolution> {
-        let mut new_goals = Vec::new();
-        self.repeat_while_none(
+    #[instrument(level = "debug", skip(self))]
+    fn set_projection_eq_hack_goal(&mut self, goal: Goal<'tcx, ty::ProjectionPredicate<'tcx>>) {
+        assert!(
+            self.nested_goals.projection_eq_hack_goal.is_none(),
+            "attempted to set the projection eq hack goal when one already exists"
+        );
+        self.nested_goals.projection_eq_hack_goal = Some(goal);
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    fn add_goal(&mut self, goal: Goal<'tcx, ty::Predicate<'tcx>>) {
+        self.nested_goals.goals.push(goal);
+    }
+
+    #[instrument(level = "debug", skip(self, goals))]
+    fn add_goals(&mut self, goals: impl IntoIterator<Item = Goal<'tcx, ty::Predicate<'tcx>>>) {
+        let current_len = self.nested_goals.goals.len();
+        self.nested_goals.goals.extend(goals);
+        debug!("added_goals={:?}", &self.nested_goals.goals[current_len..]);
+    }
+
+    // Recursively evaluates all the goals added to this `EvalCtxt` to completion, returning
+    // the certainty of all the goals.
+    #[instrument(level = "debug", skip(self))]
+    fn try_evaluate_added_goals(&mut self) -> Result<Certainty, NoSolution> {
+        let mut goals = core::mem::replace(&mut self.nested_goals, NestedGoals::new());
+        let mut new_goals = NestedGoals::new();
+
+        let response = self.repeat_while_none(
             |_| Ok(Certainty::Maybe(MaybeCause::Overflow)),
             |this| {
                 let mut has_changed = Err(Certainty::Yes);
-                for goal in goals.drain(..) {
-                    let (changed, certainty) = match this.evaluate_goal(goal) {
+
+                if let Some(goal) = goals.projection_eq_hack_goal.take() {
+                    let (_, certainty) = match this.evaluate_goal(
+                        true,
```

Comment:
please make this a custom enum :thinking: `in_projection_eq` hack is too weird to be a `bool`.

also do you think it makes sense to 1) move this whole `if` into a separate private function, something like `evaluate_projection_eq_hack_goal` and 2) move `try_evaluate_added_goals` into a separate module together with `evaluate_goal`? :thinking:

Reply:
made it an enum, 

I don't think it makes much sense to move this `if` to a separate fn, most of the logic here is specific to rerunning things in the loop and not something that i would expect to see in a `evaluate_projection_eq_hack_goal`. 

Moving `try_evaluate_added_goals` into a separate module would be good I think, I thought it was a little odd that we had `evaluate_goal` in the `solve/mod.rs` file instead of elsewhere, if we move `evaluate_goal` we can also maybe make it private so that you cant call it instead of `add_goal` which would be good I think. Would you be fine with this happening in a different PR or do you want it to be done now

Reply:
would prefer it done now, not strictly required though :shrug: